src/snowflake/snowpark/mock/mock_plan.py

```diff
@@ -254,6 +279,149 @@ def execute_mock_plan(plan: MockExecutionPlan) -> TableEmulator:
+            on, BinaryExpression
+        ):  # ON a condition, apply where to a Cartesian product
+            on = None
+        else:  # ON clause not specified, SF returns a Cartesian product
+            on = None
+
+        # Processing the join type
+        how = source_plan.join_type.sql
+        if how.startswith("USING "):
+            how = how[6:]
+        if how.startswith("NATURAL "):
+            how = how[8:]
+        if how == "LEFT OUTER":
+            how = "LEFT"
+        elif how == "RIGHT OUTER":
+            how = "RIGHT"
+        elif "FULL" in how:
+            how = "OUTER"
+        elif "SEMI" in how:
+            how = "INNER"
+        elif "ANTI" in how:
+            how = "CROSS"
+
+        if (
+            "NATURAL" in source_plan.join_type.sql and on is None
+        ):  # natural joins use the list of common names as keys
+            on = left.columns.intersection(right.columns).values.tolist()
+
+        if on is None:
+            how = "CROSS"
+
+        result_df = left.merge(
+            right,
+            on=on,
+            how=how.lower(),
+        )
+
+        # Restore sf_types information after merging, there should be better way to do this
+        result_df.sf_types.update(left.sf_types)
+        result_df.sf_types.update(right.sf_types)
+
+        if on:
+            result_df = result_df.reset_index(drop=True)
+            if isinstance(on, list):
+                # Reorder columns for JOINS with USING clause, where Snowflake puts the key columns to the left
+                reordered_cols = on + [
+                    col for col in result_df.columns.tolist() if col not in on
+                ]
+                result_df = result_df[reordered_cols]
+
+        common_columns = set(L_expr_to_alias.keys()).intersection(
+            R_expr_to_alias.keys()
+        )
+        new_expr_to_alias = {
+            k: v
+            for k, v in {
+                **L_expr_to_alias,
+                **R_expr_to_alias,
+            }.items()
+            if k not in common_columns
+        }
+        expr_to_alias.update(new_expr_to_alias)
+
+        if source_plan.condition:
+            condition = calculate_expression(
+                source_plan.condition, result_df, analyzer, expr_to_alias
+            )
+
+            if "SEMI" in source_plan.join_type.sql:  # left semi
+                result_df = left[
+                    left.apply(tuple, 1).isin(
+                        result_df[condition][left.columns].apply(tuple, 1)
+                    )
+                ].dropna()
+            elif "ANTI" in source_plan.join_type.sql:  # left anti
+                result_df = left[
+                    ~(
+                        left.apply(tuple, 1).isin(
+                            result_df[condition][left.columns].apply(tuple, 1)
+                        )
+                    )
+                ].dropna()
+            elif "LEFT" in source_plan.join_type.sql:  # left outer join
```

Comment:
The left outer, righter outer and full outer have some duplicate code. Refactoring...